### ISSUE TITLE
What we need to implement

### ISSUE THREAD

**Issue by ozora-ogino** [ozora-ogino/wav2spk#2]
- [ ] Temporal gating
- [ ] AM Softmax
- [ ] StatPooling

**Comment by ma7555** [comment]
@ozora-ogino,
So by temporal gating, the author is referring to Squeeze and Excitation, which as quoted from [here](https://paperswithcode.com/method/squeeze-and-excitation-block) is:

- The block has a convolutional block as an input.
- Each channel is "squeezed" into a single numeric value using average pooling.
- A dense layer followed by a ReLU adds non-linearity and output channel complexity is reduced by a ratio.
- Another dense layer followed by a sigmoid gives each channel a smooth gating function.
- Finally, we weight each feature map of the convolutional block based on the side network; the "excitation".

I will attempt to deploy it.

**Comment by ma7555** [comment]
@ozora-ogino, here is the temporal gating block as per my understanding
```
def squeeze_excite_block(input_tensor, ratio=16):
    channel_axis = -1
    filters = input_tensor.shape[channel_axis]
    se_shape = (1, filters)
    se = tf.keras.layers.GlobalAveragePooling1D()(input_tensor)
    se = tf.keras.layers.Reshape(se_shape)(se)
    se = tf.keras.layers.Dense(filters // ratio, activation='relu', kernel_initializer='he_normal', use_bias=False)(se)
    se = tf.keras.layers.Dense(filters, activation='sigmoid', kernel_initializer='he_normal', use_bias=False)(se)

    x = tf.keras.layers.Multiply()([input_tensor, se])
    return x
```

**Comment by ozora-ogino** [comment]
@ma7555 
Thank you for sharing.
Your implementation looks correct.

Let me check if it works.

Thank you.

**Comment by ozora-ogino** [comment]
@ma7555, I implemented temporal gating as a subclassed layer referring to your code.
This is better than lambda layer.

```
class TemporalGating(tf.keras.layers.Layer):
 def __init__(self, ration=16, data_format='channel_last'):
  super(TemporalGating, self).__init__()
  self.data_format = data_format
  self.ratio = ratio
  self. pooling = tf.keras.layers.GlobalAveragePooling1D()
  self.elu_dense = None
  self.linear_dense = None


def call(self, inputs):
  x = self.pooling(inputs)
  if self.data_format is 'channel_last:
    x = tf.keras.layers.Reshape((1, inputs.shape[-1]))(x)
  else:
    x = tf.keras.layers.Reshape((1, inputs.shape[1]))(x)
  if not self.elu_dense:
    self.elu_dense = tf.keras.layers.Dense(inputs.shape[-1] // self.ratio, activation='relu', kernel_initializer='he_normal', use_bias=False)
  x = self.elu_dense(x)
  if not self.linear_dense:
    self.linear_dense = tf.keras.layers.Dense(inputs.shape[-1], activation='sigmoid', kernel_initializer='he_normal', use_bias=False)
  x = self.linear_dense(x)
  x = tf.keras.layers.Multiply()([inputs, x])
  return x
```

**Comment by ozora-ogino** [comment]
Hi @ma7555,
How is it going?

I implemented AM-Softmax which can be used with TF2.0.
Please check my repository.
https://github.com/ozora-ogino/asoftmax-tf

**Comment by ma7555** [comment]
@ozora-ogino

Hey pal, everything is going well.

I will check give it a try this week and maybe look into stats pooling. Maybe we can cooperate on it.

**Comment by ozora-ogino** [comment]
@ma7555 ,

Okay.
Do you understand about stats pooling?
